x/collection/keeper/msg_server_test.go

```diff
@@ -12,122 +14,190 @@ import (
-			contractID: s.contractID,
-			amount:     s.balance,
-			events:     sdk.Events{sdk.Event{Type: "lbm.collection.v1.EventSent", Attributes: []abci.EventAttribute{{Key: []uint8{0x61, 0x6d, 0x6f, 0x75, 0x6e, 0x74}, Value: []uint8{0x5b, 0x7b, 0x22, 0x74, 0x6f, 0x6b, 0x65, 0x6e, 0x5f, 0x69, 0x64, 0x22, 0x3a, 0x22, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x31, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x22, 0x2c, 0x22, 0x61, 0x6d, 0x6f, 0x75, 0x6e, 0x74, 0x22, 0x3a, 0x22, 0x31, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x22, 0x7d, 0x5d}, Index: false}, {Key: []uint8{0x63, 0x6f, 0x6e, 0x74, 0x72, 0x61, 0x63, 0x74, 0x5f, 0x69, 0x64}, Value: []uint8{0x22, 0x39, 0x62, 0x65, 0x31, 0x37, 0x31, 0x36, 0x35, 0x22}, Index: false}, {Key: []uint8{0x66, 0x72, 0x6f, 0x6d}, Value: []uint8{0x22, 0x6c, 0x69, 0x6e, 0x6b, 0x31, 0x76, 0x39, 0x6a, 0x78, 0x67, 0x75, 0x6e, 0x39, 0x77, 0x64, 0x65, 0x6e, 0x71, 0x61, 0x32, 0x78, 0x7a, 0x66, 0x78, 0x22}, Index: false}, {Key: []uint8{0x6f, 0x70, 0x65, 0x72, 0x61, 0x74, 0x6f, 0x72}, Value: []uint8{0x22, 0x6c, 0x69, 0x6e, 0x6b, 0x31, 0x76, 0x39, 0x6a, 0x78, 0x67, 0x75, 0x6e, 0x39, 0x77, 0x64, 0x65, 0x6e, 0x71, 0x61, 0x32, 0x78, 0x7a, 0x66, 0x78, 0x22}, Index: false}, {Key: []uint8{0x74, 0x6f}, Value: []uint8{0x22, 0x6c, 0x69, 0x6e, 0x6b, 0x31, 0x76, 0x39, 0x6a, 0x78, 0x67, 0x75, 0x6e, 0x39, 0x77, 0x64, 0x65, 0x6e, 0x79, 0x6a, 0x71, 0x79, 0x79, 0x78, 0x75, 0x22}, Index: false}}}},
+			req: &collection.MsgSendFT{
+				ContractId: s.contractID,
+				From:       s.vendor.String(),
+				To:         s.customer.String(),
+				Amount:     collection.NewCoins(collection.NewFTCoin(s.ftClassID, s.balance)),
+			},
+			ftID: collection.NewFTID(s.ftClassID),
+			expectedEvent: helperBuildEventSent(s.contractID, s.vendor, s.customer, s.vendor, collection.NewCoins(
+				collection.NewFTCoin(s.ftClassID, s.balance),
+			)),
 		},
 		"contract not found": {
-			contractID: "deadbeef",
-			amount:     s.balance,
-			err:        class.ErrContractNotExist,
+			isNegativeCase: true,
+			req: &collection.MsgSendFT{
+				ContractId: "deadbeef",
+				From:       s.vendor.String(),
+				To:         s.customer.String(),
+				Amount:     collection.NewCoins(collection.NewFTCoin(s.ftClassID, s.balance)),
+			},
+			ftID:          collection.NewFTID(s.ftClassID),
+			expectedError: class.ErrContractNotExist,
 		},
 		"insufficient funds": {
-			contractID: s.contractID,
-			amount:     s.balance.Add(sdk.OneInt()),
-			err:        collection.ErrInsufficientToken,
+			isNegativeCase: true,
+			req: &collection.MsgSendFT{
+				ContractId: s.contractID,
+				From:       s.vendor.String(),
+				To:         s.customer.String(),
+				Amount:     collection.NewCoins(collection.NewFTCoin(s.ftClassID, s.balance.Add(sdk.OneInt()))),
+			},
+			ftID:          collection.NewFTID(s.ftClassID),
+			expectedError: collection.ErrInsufficientToken,
 		},
 	}
 
 	for name, tc := range testCases {
 		s.Run(name, func() {
-			ctx, _ := s.ctx.CacheContext()
```

Comment:
Is it OK to stop using the cache? Generally, it will make other tests affected by it, which often means, making a race condition. Of course, in this case, it may not be the case.

Reply:
I'll look at it again.